quickannotator/api/v1/utils/shared_crud.py

```diff
@@ -58,44 +59,132 @@ def get_annotation_query(model, scale_factor: float=1.0) -> Query:
+        'gt_datetime': datetime.now()
+    }
+    
+    stmt = insert(models.Tile).values(
+        [
+            {
+                'annotation_class_id': annotation_class_id,
+                'image_id': image_id,
+                'tile_id': tile_id,
+                **update_fields
+            }
+            for tile_id in tile_ids
+        ]
+    ).on_conflict_do_update(
         index_elements=['annotation_class_id', 'image_id', 'tile_id'],
         set_=update_fields
+    ).returning(
+        *[getattr(models.Tile, column.name) for column in models.Tile.__table__.columns]   # Can't use models.Tile with on_conflict_do_update
     )
 
-    result = db_session.execute(stmt)
+    tiles = db_session.execute(stmt).all()
     db_session.commit()
+    return tiles
+
 
-    return result
+
+def upsert_pred_tiles(image_id: int, annotation_class_id: int, tile_ids: List[int], pred_status: TileStatus, process_owns_tile=False) -> List[models.Tile]:
+    """
+    Inserts new prediction tiles or updates existing prediction tiles in the database.
+    This function attempts to insert new prediction tiles with the given annotation_class_id, image_id, and tile_ids.
+    If prediction tiles with the same annotation_class_id, image_id, and tile_ids already exist, it updates the relevant fields.
+
+    Args:
+        image_id (int): The ID of the image.
+        annotation_class_id (int): The ID of the annotation class.
+        tile_ids (List[int]): The IDs of the tiles.
+        pred_status (TileStatus): The status of the tile prediction.
+        process_owns_tile (bool, optional): If True, updates tiles regardless of their current status. Defaults to False. Only a ray actor should set this to True.
+
+    Returns:
+        List[models.Tile]: The tiles that were inserted or updated.
+    """
+
+    if len(tile_ids) == 0:
+        return []
+
+    if pred_status is None:
+        raise ValueError("pred_status cannot be None.")
+    
+    update_fields = {
+        'pred_status': pred_status,
+        'pred_datetime': datetime.now()
+    }
+    
+    if process_owns_tile:
+        set = update_fields
+    else:
+        current_time = datetime.now()
+        expiration_thresh = current_time - timedelta(minutes=constants.TILE_PRED_EXPIRE)
+        set = {
+            'pred_status': sqlalchemy.case(
+                (models.Tile.pred_status == TileStatus.UNSEEN, pred_status),
+                (
+                    (models.Tile.pred_status == TileStatus.DONEPROCESSING) & 
+                    (models.Tile.pred_datetime <= expiration_thresh), pred_status
+                ),
+            else_=models.Tile.pred_status,
+            ),
+            'pred_datetime': sqlalchemy.case(
+                (models.Tile.pred_status == TileStatus.UNSEEN, datetime.now()),
+                (
+                    (models.Tile.pred_status == TileStatus.DONEPROCESSING) & 
+                    (models.Tile.pred_datetime <= expiration_thresh), datetime.now()
+                ),
```

Comment:
this makes me feel a bit uncomfortable because it'd take me a lot of brain power to parse and check : ) are you sure this is (a) correct and (b) the easiest way to do this?

Reply:
1. Need on_conflict_do_update to keep the operation atomic
2. We didn't use case logic originally. The previous approach used a `where` clause to filter by UNSEEN or stale DONEPROCESSING. However, the previous approach only returned tiles which satisfied the update condition, rather than returning the tile regardless of whether or not it was updated.

That said, the current method is not very readable.

I've just pushed an update which should be more readable and has less redundant code. Thoughts?

Reply:
ohhh yes - much more attractive and easy to parse, thanks